### PR TITLE
feat: add codegen utility

### PR DESCRIPTION
**Pull request by ZachJW34**
- Closes [UNIFY-446](https://cypress-io.atlassian.net/jira/software/projects/UNIFY/boards/20?selectedIssue=UNIFY-446)

### Details
This PR adds the ability to generate files from ejs templates as well as normal text/binary files. Given a directory, the `codeGenerator` function will recursively grab all files within the specified directory and move them to the specified destination whilst preserving the folder structure. Any values passed will be interpolated into the files if the file is using standard [ejs](https://ejs.co/) templating syntax. Files also support the use of [front-matter](https://www.npmjs.com/package/front-matter) to specify custom filenames that aren't static. An example use case for the Cypress config case would look like:
```
---
fileName: cypress.config.<%=lang%>
---

export default defineConfig(...)
```

This PR also replaces the current solution for generating specs from component/story/integration. The data-model for handling the generated files via graphql can be improved but I thought I'd keep the current model for now since there is active work ongoing with the current graphql endpoints involving spec generation. 

The video below shows the result of running `yarn workspace @packages/data-context test-unit`. The given templateDir is `template-test` and target is `tmp/test-code-gen/test`.

https://user-images.githubusercontent.com/25158820/139487151-676a1f97-9cbb-4e74-859d-a19411b86be8.mov

### Testing
The `packages/data-context/test/unit/code-generator.spec.ts` test checks the output of the code-generator and also parses the `story, component, integration` templates to verify that we are producing valid javascript given valid arguments. The `packages/app/cypress/e2e/integration/new-spec.spec.ts` e2e test also generates from these templates given a real world app. This flow can also be tested by running `yarn dev` and navigating to `__vite__/#/newspec` in the runner.

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/src/codegen/spec-options.ts [hunk 1]
```diff
@@ -167,36 +128,37 @@ export class SpecGenerator {
...
+        fileName: this.getFilename(this.parsedPath.ext),
       },
       vue: {
         imports: [
           'import { mount } from "@cypress/vue"',
           'import { composeStories } from "@storybook/testing-vue3"',
           `import * as stories from "./${this.parsedPath.name}"`,
         ],
-        mount: (StoryName: string) => `mount(${StoryName}())`,
+        stories: csf.stories.map((story) => {
```
Comment (lmiller1990) [inline packages/data-context/src/codegen/spec-options.ts:147]: Will the above `imports` eventually move to a codegen file as well? (as opposed to inlined here?)

File: packages/data-context/src/codegen/code-generator.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  args: { [key: string]: any },
+): Promise<CodeGenResults> {
+  const templateFiles = await allFilesInDir(action.templateDir)
+  const codeGenResults: CodeGenResults = { files: [], failed: [] }
+
+  for (const file of templateFiles) {
+    const isBinary = await isBinaryFile(file)
+    const parsedFile = path.parse(file)
+
+    const processBinaryFile = async () => {
```
Comment (lmiller1990) [inline packages/data-context/src/codegen/code-generator.ts:43]: What is the use case for this? Are we ever generated a binary file? What kind of file?

**Review by lmiller1990 — COMMENTED** [review]
Left a few questions.

The actual infrastructure, which is the point of this PR, seems to work fine - I tried it out.

The actual codegen snippets need work - guessing that is outside the scope of this PR? For example I generated this for the launchpad project:

```
import { mount } from "@cypress/vue"
import App from "./App.vue"

describe('<App />', () => {
  it('renders', () => {
    // see: https://reactjs.org/docs/test-utils.html
    mount(App, { props: {} })
  })
})
```

Several weird things:
- empty props object doesn't seem very useful
- has link to react docs for my Vue component?

Might be good to make an issue to go over the snippets. Having snapshots for these would help greatly, since we wouldn't need to generate them manually to test them.

Do you plan to add snapshot tests as well (either here or later?)

File: packages/data-context/src/codegen/spec-options.ts (hunk 1 above)
Comment (JessicaSachs) [reply packages/data-context/src/codegen/spec-options.ts:147]: They could be. Right now we're not looping within ejs.

File: packages/data-context/src/codegen/code-generator.ts (hunk 2 above)
Comment (JessicaSachs) [reply packages/data-context/src/codegen/code-generator.ts:43]: pngs

File: packages/data-context/src/codegen/spec-options.ts (hunk 1 above)
Comment (ZachJW34) [reply packages/data-context/src/codegen/spec-options.ts:147]: I was mostly wiring it up to the old way of generating specs. I agree, a lot of the options could and should be moved into the template files. We could even have a `vue-component` and a `react-component` template if we didn't want to muddy the templates with too much logic. I'll create an issue for cleaning up the templates.

File: packages/data-context/src/codegen/code-generator.ts (hunk 2 above)
Comment (ZachJW34) [reply packages/data-context/src/codegen/code-generator.ts:43]: I took a lot of inspiration from [node-plop](https://github.com/plopjs/node-plop/blob/22aa15e31026d48ddb9a29419597b8fec2d78b02/src/actions/_common-action-add-file.js#L37) and [nx](https://github.com/nrwl/nx/blob/27cc2d9c5f1e90ee4246a314c13b6c87fb7ac026/packages/devkit/src/generators/generate-files.ts#L6). They both had special handling for binary files and it wasn't too difficult to include it so I included it! Might be useful in the future.

**Comment by ZachJW34** [comment]
@lmiller1990 Fixed the vue docs link. Snapshot testing and cleaner templates will be nice so I'll create an issue for it.

**Comment by ZachJW34** [comment]
Looks like #18714 will fix the failing `create-build-artifacts` step.

**Pull request merged**